src/relations/osm_mysql.py

```diff
@@ -0,0 +1,153 @@
+        super().__init__(charm, LEGACY_OSM_MYSQL)
+
+        self.charm = charm
+
+        self.framework.observe(self.charm.on.leader_elected, self._on_leader_elected)
+        self.framework.observe(
+            self.charm.on[LEGACY_OSM_MYSQL].relation_created, self._on_osm_mysql_relation_created
+        )
+        self.framework.observe(
+            self.charm.on[LEGACY_OSM_MYSQL].relation_broken, self._on_osm_mysql_relation_broken
+        )
+
+    def _get_or_set_password_in_peer_databag(self, username: str) -> str:
+        """Get a user's password from the peer databag if it exists, else populate a password.
+
+        Args:
+            username: The mysql username
+
+        Returns:
+            a string representing the password for the mysql user
+        """
+        peer_databag = self.charm._peers.data[self.charm.app]
+
+        if peer_databag.get(f"{username}_password"):
+            return peer_databag.get(f"{username}_password")
+
+        password = generate_random_password(PASSWORD_LENGTH)
+        peer_databag[f"{username}_password"] = password
+
+        return password
+
+    def _on_leader_elected(self, _) -> None:
+        """Handle the leader elected event.
+
+        Retrieves relation data from the peer relation databag and copies
+        the relation data into the new leader unit's databag.
+        """
+        # Skip if the charm is not past the setup phase (config-changed event not executed yet)
+        if not self.charm._is_peer_data_set:
+            return
+
+        relation_data = json.loads(
+            self.charm._peers.data[self.charm.app].get("osm_mysql_relation_data", "{}")
+        )
+
+        for relation in self.charm.model.relations.get(LEGACY_OSM_MYSQL, []):
+            relation_databag = relation.data
+
+            # Copy relation data into the new leader unit's databag
+            for key, value in relation_data.items():
+                if relation_databag[self.charm.unit].get(key) != value:
+                    relation_databag[self.charm.unit][key] = value
+
+            # Assign the cluster primary's address as the database host
+            primary_address = self.charm._mysql.get_cluster_primary_address().split(":")[0]
+            relation_databag[self.charm.unit]["host"] = primary_address
+
+    def _on_osm_mysql_relation_created(self, event: RelationCreatedEvent) -> None:
+        """Handle the legacy 'mysql' relation created event.
+
+        Will set up the database and the scoped application user. The connection
+        data (relation data) is then copied into the peer relation databag (to
+        be copied over to the new leader unit's databag in case of a new leader
+        being elected).
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        # Wait until on-config-changed event is executed
+        # (wait for root password to have been set)
+        if not self.charm._is_peer_data_set:
+            event.defer()
+            return
```

Comment:
In order `self.charm._is_peer_data_set` the  ` on-config-changed ` event needs to run deferring here can cause this event to continuously defer making it take much longer for the config-changed hook to get on the queue before this event re-queues. This can lead to downtime/poor performance. Instead I suggest you set a flag and trigger and event that will re-call `_on_osm_mysql_relation_created`.

Mykola did something similar in his k8s charm [here](https://github.com/delgod/mongodb-k8s-operator/blob/ad178a7e2406f3b805ac73d5dfb7927b123c6deb/lib/charms/mongodb_libs/v0/mongodb_provider.py#L60-L63) and instead of re-triggering the event, he creating a function for himself to call once his condition was met by the start hook, see [here](https://github.com/delgod/mongodb-k8s-operator/blob/ad178a7e2406f3b805ac73d5dfb7927b123c6deb/src/charm.py#L142)

I think Paulo mentioned that this delay will only occur in a bundle but I think its worth looking out for.

Reply:
I'm not sure that this idea will work in this case. There are 2 conditions that must be met before the event handler (function) can be called:
1. The `on-config-changed` should have completed execution at least once
2. The `osm-mysql` relation must have been created
With these conditions and machinery at hand, I believe that the existing approach is sufficient (and I agree with Paulo that this will only happen once in a bundle at the time of relation being created)